smartsim/_core/_install/platform.py

```diff
@@ -0,0 +1,241 @@
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import enum
+import json
+import os
+import pathlib
+import platform
+import typing as t
+from dataclasses import dataclass
+
+from typing_extensions import Self
+
+
+class PlatformError(Exception):
+    pass
+
+
+class UnsupportedError(PlatformError):
+    pass
+
+
+class Architecture(enum.Enum):
+    """Identifiers for supported CPU architectures
+
+    :raises UnsupportedError: If a given architecture is not enumerated
+                              (i.e not supported)
+    :return: An enum representing the CPU architecture
+    """
+
+    X64 = "x86_64"
+    ARM64 = "arm64"
+
+    @classmethod
+    def from_str(cls, string: str) -> "Architecture":
+        """Return enum associated with the architecture
+
+        :param string: String representing the architecture, see platform.machine
+        :raises UnsupportedError: If not enumerated here, unsupported architecture
+        :return: Enum for a specific architecture
+        """
+        string = string.lower()
+        for type_ in cls:
+            if string in type_.value:
+                return type_
+        raise UnsupportedError(f"Unrecognized or unsupported architecture: {string}")
+
+    @classmethod
+    def autodetect(cls) -> "Architecture":
+        """Automatically return the architecture of the current machine
+
+        :return: enum of this platform's architecture
+        """
+        return cls.from_str(platform.machine())
+
+
+class Device(enum.Enum):
+    """Identifiers for the device stack
+
+    :return: Enum associated with the device stack
+    """
+
+    CPU = "cpu"
+    CUDA118 = "cuda-11.8"
+    CUDA121 = "cuda-12.1"
+    ROCM57 = "rocm-5.7"
+
+    @classmethod
+    def from_str(cls, str_: str) -> "Device":
```

Comment:
just wanna make sure you want this return type not to be optional, since it is optional elsewhere.

Reply:
Yes, this should always return something valid (or throw an Exception)